helix-gateway/src/main/proto/HelixGatewayService.proto

```diff
@@ -0,0 +1,52 @@
+syntax = "proto3";
+
+package proto.org.apache.helix.gateway;
+
+message SingleTransitionMessage {
+  enum TransitionType {
+    ADD_SHARD = 0;
+    DELETE_SHARD = 1;
+    CHANGE_ROLE = 2;
+  }
+  string transitionID = 1;    // ID of transition message
+  TransitionType transitionType = 2;   // Transition type for shard operations
+  string resourceID = 3;              // Resource ID
+  string shardID = 4;                 // Shard to perform operation
+  string startState = 5;      // Shard start state
+  string targetState = 6;     // Shard target state.
+}
+
+message StateTransitionMessage{
+  repeated SingleTransitionMessage request = 1;
+}
+
+message SingleReplicaStateTransitionStatus {
```

Comment:
Shall we ignore replica in the naming? It is asymmetrical for naming in request/response. Also it is status of message also means for a replica already.

Reply:
The outer request is names "replicaState", so I think probably we want to keep the replica in name. But we can simplify it as "SingleReplicaTransitionStatus"